modules/cudaarithm/src/cuda/threshold.cu

```diff
@@ -95,12 +95,232 @@ namespace
     }
 }
 
-double cv::cuda::threshold(InputArray _src, OutputArray _dst, double thresh, double maxVal, int type, Stream& stream)
+template <uint n_bins>
+__global__ void otsu_mean(uint *histogram, uint *threshold_sums, unsigned long long *sums)
+{
+    extern __shared__ unsigned long long shared_memory_u64[];
+
+    uint bin_idx = blockIdx.x * blockDim.x + threadIdx.x;
```

Comment:
This is confusing.  As you have 256 threads per block and 256 blocks I would suggest the following launch parameters
```
dim3 block_all = dim3(n_bins);
dim3 grid_all = dim3(n_thresholds);
dim3 block_score = dim3(n_thresholds);
dim3 grid_score = dim3(1);
```
with
```
uint bin_idx = threadIdx.x;
uint threshold = blockIdx.x;
```
`int`is usually more performant than `uint`.